contrib/sly-mrepl.el

```diff
@@ -1178,24 +1178,22 @@ Doesn't clear input history."
 (defun sly-mrepl-next-prompt ()
   "Go to the beginning of the next REPL prompt."
   (interactive)
-  (let ((pos (next-single-char-property-change (line-beginning-position 2)
-                                               'sly-mrepl--prompt)))
-    (goto-char pos))
+  (let ((at-prompt-p (get-text-property (point) 'sly-mrepl--prompt)))
+    (dotimes (_ (if at-prompt-p 1 2))
+      (goto-char (next-single-char-property-change (point) 'sly-mrepl--prompt))))
   (end-of-line))
 
 (defun sly-mrepl-previous-prompt ()
   "Go to the beginning of the previous REPL prompt."
   (interactive)
-  ;; This has two wrinkles around the first prompt: (1) when going to
-  ;; the first prompt it leaves point at column 0 (1) when called from
-  ;; frist prompt goes to beginning of buffer.  The correct fix is to
-  ;; patch comint.el's comint-next-prompt and comint-previous-prompt
-  ;; anyway...
-  (let* ((inhibit-field-text-motion t)
-         (pos (previous-single-char-property-change (1- (line-beginning-position))
-                                                   'sly-mrepl--prompt)))
-    (goto-char pos)
-    (goto-char (line-beginning-position)))
+  (let ((at-prompt (let ((inhibit-field-text-motion t))
+                     (get-text-property (line-beginning-position) 'sly-mrepl--prompt))))
+    (when at-prompt
+      (let ((inhibit-field-text-motion t))
+        (beginning-of-line)
+        (forward-char)))
+    (dotimes (_ (if at-prompt 2 1))
+      (goto-char (previous-single-char-property-change (point) 'sly-mrepl--prompt))))
```

Comment:
Right this looks much better, but I think what you're looking for is not a brittle 2/1 repetition factor but rather looking for a specific **value** of the property 'sly-mrepl-prompt, which is where you should stop.  `cl-loop`, Elisp's version of CL `loop` is probably your friend here.

Reply:
The `dotimes` is simply to avoid repeating the `(goto-char ...)` twice.
There is no need for looping here.

Without `dotimes`, the function would look like this:

```lisp
(defun sly-mrepl-previous-prompt ()
  "Go to the beginning of the previous REPL prompt."
  (interactive)
  (let ((at-prompt (let ((inhibit-field-text-motion t))
                     (get-text-property (line-beginning-position) 'sly-mrepl--prompt))))
    (when at-prompt
      (let ((inhibit-field-text-motion t))
        (beginning-of-line)
        (forward-char)
        (goto-char (previous-single-char-property-change (point) 'sly-mrepl--prompt))))
    (goto-char (previous-single-char-property-change (point) 'sly-mrepl--prompt)))
  (end-of-line))
```

Reply:
I'm not suggesting you do `goto-char` more than needed.  Tip: use `text-properties-at`

Reply:
Alternatively, I can `flet` the `goto-char`, maybe it's more readable:

```lisp
(defun sly-mrepl-previous-prompt ()
  "Go to the beginning of the previous REPL prompt."
  (interactive)
  (cl-flet ((at-prompt-p ()
                         (let ((inhibit-field-text-motion t))
                           (get-text-property (line-beginning-position) 'sly-mrepl--prompt)))
            (move-back ()
                       (goto-char (previous-single-char-property-change (point) 'sly-mrepl--prompt))))
    (when (at-prompt-p)
      (let ((inhibit-field-text-motion t))
        (beginning-of-line)
        (forward-char)
        (move-back)))
    (move-back))
  (end-of-line))
```

Reply:
Also looping is useful if you wnat to support an optional argument `n` that tells you how many prompts back (or forward) you wish to navigate.  This is consistent with many other like-minded Emacs commands.

Reply:
Basically I want:

```
(defun sly-mrepl-previous-prompt (n)
  "Go to beginning of Nth previous prompt."
...)
```
It should have a nice interactive spec so that, say, `M-2 M-x sly-mrepl-previous-prompt` goes back 2 prompts.

Reply:
OK for the `n` argument.

> I'm not suggesting you do goto-char more than needed. Tip: use text-properties-at

I don't understand.  The text property is always the same for all prompts.  How would text-properties-at help here?

Reply:
Also I'm calling `goto-char` exactly the requirement number of times I believe.

Reply:
> Also I'm calling goto-char exactly the requirement number of times I believe.

I don't dispute that, it's just that the `dotimes` sometimes 1 sometimes 2 is a code smell.  When you implement the `n` arg you will see.  Bonus points if you can make the `backward` use the `forward` or vice versa.

Reply:
So what do you think of this:

```lisp
(defun sly-mrepl-previous-prompt (&optional count)
  "Go to the beginning of the previous REPL prompt."
  (interactive "p")
  (setf count (or count 1))
  (cl-flet ((at-prompt-p ()
                         (let ((inhibit-field-text-motion t))
                           (get-text-property (line-beginning-position) 'sly-mrepl--prompt)))
            (move-back ()
                       (goto-char (previous-single-char-property-change (point) 'sly-mrepl--prompt))))
    (dotimes (_ count)
      (when (at-prompt-p)
        (let ((inhibit-field-text-motion t))
          (beginning-of-line))
        (forward-char)
        (move-back))
      (move-back)))
  (end-of-line))
```

I tried having `backward` reuse forward, but there is too much code specialization for each function, having a single function results in code that's as long, if not longer, and harder to understand.